# Task 1 - Data Analysis

In [108]:
import numpy as np
import pandas as pd

In [109]:
data = pd.read_csv('/content/raw_data_v11_3.csv')

In [110]:
data.head()

Date of Birth                   What is your current profession?  \
0     9/16/2000  Fresher (I have completed my Education and loo...   
1      2/6/1998  Fresher (I have completed my Education and loo...   
2     10/8/2003  Student (I am currently enrolled in some Educa...   
3    11/29/2001  Student (I am currently enrolled in some Educa...   
4    11/25/2004  Student (I am currently enrolled in some Educa...   

  What is the year of completion of your Highest Education Qualification?  \
0                                          7/30/2023                        
1                                          5/18/2022                        
2                                           4/1/2025                        
3                                          5/15/2024                        
4                                          1/10/2026                        

          Which of the following best describes you?  \
0  I am currently enrolled in a Data Science cert...   
1  I have never enrolled in any certification cou...   
2  I am currently enrolled in a Data Science cert...   
3  I have never enrolled in any certification cou...   
4  I have completed a Data Science certification ...   

  Please mention the name of the institute where you have completed or are currently pursuing your Data Science Certification Course.  \
0                           Innomatics Research Labs                                                                                    
1                                                NaN                                                                                    
2                                                NaN                                                                                    
3                                                NaN                                                                                    
4         Velagapudi ramakrishna Siddhartha college                                                                                     

  In case your certification institute is "Innomatics Research Labs", please provide your Enrolled Batch Number.  \
0                                                251                                                               
1                                                NaN                                                               
2                                                NaN                                                               
3                                                NaN                                                               
4                                                 Na                                                               

  Please indicate the time that suits you for your daily LIVE interaction with our Chief Data Scientist as part of your internship.  \
0                                6:00 to 7:00 PM IST                                                                                  
1                                6:00 to 7:00 PM IST                                                                                  
2                                6:00 to 7:00 PM IST                                                                                  
3                                5:00 to 6:00 PM IST                                                                                  
4                                6:00 to 7:00 PM IST                                                                                  

  What do you think are most important skill right now for a skilled Data Scientist?  
0  Extracting actionable insights from the raw da...                                  
1   Extracting actionable insights from the raw data                                  
2                                     Data Modelling                                  
3  Extracting actionable insights from the raw da...                                  
4  Extracting a

Renaming columns for easy usage

In [111]:
data.rename(columns = {data.columns[0]:'DOB',
                       data.columns[1]:'Profession',
                       data.columns[2]:'YOG',
                       data.columns[3]:'Self_Description',
                       data.columns[4]:'DS_Course',
                       data.columns[5]:'Innomatics_batch_no',
                       data.columns[6]:'session_timings',
                       data.columns[7]:'DS_skills'}, inplace = True)
data.describe()

DOB                                         Profession  \
count         1000                                               1000   
unique         826                                                  4   
top     11/24/2001  Student (I am currently enrolled in some Educa...   
freq             4                                                488   

              YOG                                   Self_Description  \
count        1000                                               1000   
unique        553                                                  3   
top     5/15/2024  I am currently enrolled in a Data Science cert...   
freq           19                                                405   

                       DS_Course Innomatics_batch_no      session_timings  \
count                        678                 338                 1000   
unique                       365                 139                    2   
top     Innomatics Research Labs                  Na  6:00 to 7:00 PM IST   
freq                         109                  16                  672   

                                                DS_skills  
count                                                1000  
unique                                                 60  
top     Extracting actionable insights from the raw da...  
freq                                                  214

In [112]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   DOB                  1000 non-null   object
 1   Profession           1000 non-null   object
 2   YOG                  1000 non-null   object
 3   Self_Description     1000 non-null   object
 4   DS_Course            678 non-null    object
 5   Innomatics_batch_no  338 non-null    object
 6   session_timings      1000 non-null   object
 7   DS_skills            1000 non-null   object
dtypes: object(8)
memory usage: 62.6+ KB


**Question 1:**

In [113]:
data['Profession'].value_counts()

Student (I am currently enrolled in some Educational Program)                          488
Fresher (I have completed my Education and looking for a job now)                      424
Working Professional (I have completed my Education and working full time)              62
Freelancer (I have completed my Education and working part time or owns a business)     26
Name: Profession, dtype: int64

**Ans:** <br>
The above data shows that most of the interns are students who are still studying. And freshers are 2nd highest very close in number too

**Question 2:**

In [114]:
print(data['session_timings'].unique())
print(data['session_timings'].value_counts())

['6:00 to 7:00 PM IST' '5:00 to 6:00 PM IST']
6:00 to 7:00 PM IST    672
5:00 to 6:00 PM IST    328
Name: session_timings, dtype: int64


**Ans:** <br> 1). 6 - 7 PM IST and 5 - 6 PM IST are the preferred timings given by the interns for live sessions <br> 2). 6-7 PM being the most preferred


**Question 3:**

In [115]:
data.shape

(1000, 8)

In [116]:
print(len(data['DOB'].unique()))

#(data['DOB'].unique().value_counts() > 1)

#data['DOB'].value_counts()>1

826


**Ans:** <br> As per the dataset, we have 1000 unique interns. On further analysis, some interns happen to share birthdays on few common days

**Question 4:**

In [117]:
data['DS_Course'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1000 entries, 0 to 999
Series name: DS_Course
Non-Null Count  Dtype 
--------------  ----- 
678 non-null    object
dtypes: object(1)
memory usage: 7.9+ KB


In [118]:
data[data['DS_Course'].isna()].Profession.value_counts()

Student (I am currently enrolled in some Educational Program)                          205
Fresher (I have completed my Education and looking for a job now)                      101
Working Professional (I have completed my Education and working full time)              12
Freelancer (I have completed my Education and working part time or owns a business)      4
Name: Profession, dtype: int64

**Ans:** <br> 1). Total of 332 interns have NaN in the course, hence they have never enrolled in a Data science course <br> 2). 205 out of 332 are students. so majority of their current profession is students who are still in university

**Question 5:**

a. What is the minimum age of an intern at Innomatics Research Labs?

In [119]:
data['DOB']


0       9/16/2000
1        2/6/1998
2       10/8/2003
3      11/29/2001
4      11/25/2004
          ...    
995     8/28/1998
996     4/22/2004
997      6/8/2000
998      6/3/2003
999     7/24/2003
Name: DOB, Length: 1000, dtype: object

In [120]:
from datetime import datetime, date
def age(born):
    born = datetime.strptime(born, "%m/%d/%Y").date()
    today = date.today()
    return today.year - born.year - ((today.month,
                                      today.day) < (born.month,
                                                    born.day))

data_age = data['DOB'].apply(age)

print(data_age.sort_values())

112    -1
536    -1
307     0
834    16
844    17
       ..
711    35
274    35
684    35
598    37
606    43
Name: DOB, Length: 1000, dtype: int64


Ans:<br>
When sorted the age in ascending order, we see negative value an zero which means wrong date of birth values given by the interns. <br> So logically, 16 is the valid minimum age of all interns.

b. How many interns were born on a Sunday?

In [121]:
def getDay(born):
    born = datetime.strptime(born, "%m/%d/%Y").date()
    ts = pd.Timestamp(born.year, born.month, born.day)
    return ts.day_of_week

data_day = data['DOB'].apply(getDay)

print(data_day.value_counts()) #pd.Timestamp(2023, 01, 28)

0    163
3    147
4    146
5    145
1    140
2    134
6    125
Name: DOB, dtype: int64


**Ans:** <br> **125** interns were born on Sunday

c. In which month do the majority of interns are born?

In [122]:
def getMonth(born):
    born = datetime.strptime(born, "%m/%d/%Y").date()
    ts = pd.Timestamp(born.year, born.month, born.day)
    return ts.month

data_day = data['DOB'].apply(getMonth)

print(data_day.value_counts())

7     103
5     102
6      97
8      93
10     92
4      87
9      81
11     78
12     76
1      66
2      64
3      61
Name: DOB, dtype: int64


Ans: <br> Majority were born in **July**

d. What is the average age of a working professional?

In [123]:
data_workage = data[data['Profession'].str.startswith('W')]
#data_age =

def workAge(born):
    born = datetime.strptime(born, "%m/%d/%Y").date()
    today = date.today()
    return today.year - born.year - ((today.month,
                                      today.day) < (born.month,
                                                    born.day))

data_age = data_workage['DOB'].apply(workAge)

print(data_age.mean())

26.112903225806452


**Ans**: <br>
Average age is **approx. 26 years 1 month** for working professionals

e. What is the maximum age of a freelancer?

In [124]:
data_freelanceage = data[data['Profession'].str.startswith('F')]

def freeAge(born):
    born = datetime.strptime(born, "%m/%d/%Y").date()
    today = date.today()
    return today.year - born.year - ((today.month,
                                      today.day) < (born.month,
                                                    born.day))

data_age = data_freelanceage['DOB'].apply(freeAge)

print(data_age.max())

43


**Ans:** <br> **43** is the Maximum age of a freelancer in this internship

**Question 6:**

In [125]:
data_skill = data['DS_skills']
data_skill

0      Extracting actionable insights from the raw da...
1       Extracting actionable insights from the raw data
2                                         Data Modelling
3      Extracting actionable insights from the raw da...
4      Extracting actionable insights from the raw da...
                             ...                        
995    Extracting actionable insights from the raw da...
996                                                 LLMs
997    Extracting actionable insights from the raw da...
998    Extracting actionable insights from the raw da...
999                      LLMs, MLOps, Prompt Engineering
Name: DS_skills, Length: 1000, dtype: object

In [126]:
list_all = data['DS_skills'].str.split(r',').dropna()
arr = []
for i in list_all:
  arr.append(i)
#list_all = np.array(list_all)
#print(list_all.shape)
# list_unique = np.unique(sum(list_all, []))
# print(len(list_unique))
#list_unique
list_all = np.array(arr, dtype=object)
list_all.ndim
#print(list_all.ndim)
#list_all
#list_all
#np.count_nonzero(list_all == "LLM")

1

In [ ]:
## Possible values/options of skills
skill_options = ["Data Modelling", "LLMs", "MLOps", "Basics of backend application development using python", "Prompt Engineering", "Extracting actionable insights from the raw data"]
#skill_options = [1,2,3,4,5]
for skill in skill_options:
  count = np.count_nonzero(data_skill == skill)
  print("{} - {}".format(skill, count))

In [128]:
dm_count, llm_count, mlop_count, bdev_count, prompt_count, ext_count = 0, 0, 0, 0, 0, 0

for skill in data_skill:
  if 'Data Modelling' in skill:
    dm_count += 1

for skill in data_skill:
  if 'LLMs' in skill:
    llm_count += 1

for skill in data_skill:
  if 'MLOps' in skill:
    mlop_count += 1

for skill in data_skill:
  if 'Basics of backend application development using python' in skill:
    bdev_count += 1

for skill in data_skill:
  if 'Prompt Engineering' in skill:
    prompt_count += 1

for skill in data_skill:
  if 'Extracting actionable insights from the raw data' in skill:
    ext_count += 1

print(dm_count)
print(llm_count)
print(mlop_count)
print(bdev_count)
print(prompt_count)
print(ext_count)

825
533
596
522
425
869


**Ans:** <br> **Extracting actionable insights** and **Data Modelling** are two skills chosen by most of the interns

Question 7:

In [129]:
data_skill = data['DS_skills']

count = 0
for skill in data_skill:
  if 'Prompt Engineering' in skill:
    count += 1
print(count)

425


**Ans:** <br> 425 interns have selected Prompt Engineering as one of the options to be most important skill

### Question 8: Institue Analysis

a. How many interns didn’t mention the name of the data science institute

In [130]:
len(data[data['DS_Course'].isna()])

322

b. How many interns mentioned Innomatics as their data science institute?

In [131]:
data['DS_Course'].str.extract(pat = '([Ii]nnomatics.)').value_counts()

Innomatics     237
innomatics       8
InnomaticsR      1
dtype: int64

**Ans**: <br> Total of **246 interns** are associated with Innomatics